wptherml/optdriver.py

```diff
@@ -0,0 +1,1204 @@
+            #get weights from model
+            params = self.get_weights()
+            V = np.array( np.array(params[2]).tolist())
+            W = np.array(list(list(np.array(params[1]))[0]), dtype = np.float64)
+            bias = np.array(params[0], dtype = np.float64)[0]
+
+            return V, W, bias
+
+
+        #following 2 methods
+        #https://github.com/tsudalab/fmbqm/blob/master/fmbqm/factorization_machine.py
+
+        def triu_mask(self, input_size, F=np.ndarray):
+            #Generate a square matrix with its upper trianguler elements being 1 and others 0.
+
+            mask = np.expand_dims(np.arange(input_size), axis=0)
+            return (np.transpose(mask) < mask) * 1.0
+
+        def VtoQ(self, V, F=np.ndarray):
+            """Calculate interaction strength by inner product of feature vectors.
+            """
+            print(V.shape)
+            #input_size = V.shape[1]
+            input_size = V.shape[0]
+
+
+            #Q = np.dot(np.transpose(V), V) # (d,d)
+            Q = np.dot(V , np.transpose(V)) # (d,d)
+            #Q =   V @ V.T
+            triu = self.triu_mask(input_size)
+            #print(triu)
+            #return Q * np.triu(np.ones((input_size,input_size)), 1)
+            return Q * triu
+            
+
+
+        def get_qubo_hamiltonian_minima(self):
+            """convert factorization machine to qubo hamiltonian and solve the model, getting one result"""
+
+            params = copy.deepcopy(self.get_weights())
+
+            V = np.array(params[2], dtype = np.float64)
+
+            W = np.array(np.array(params[1])[0], dtype = np.float64)
+
+            bias = np.array(params[0], dtype = np.float64)[0]
+
+
+            Q = V @ V.T
+
+
+            Q = self.VtoQ(V)
+
+            binary_len=int(self.n)
+            qubo_basis = []
+            for i in range(0, binary_len):
+                new_string = 'x' + str(i+1)
+                qubo_basis.append(Binary(new_string))
+
+
+
+            # define qubo Hamiltonian
+            H = 0
+            for i in range(0, len(Q)):
+                for j in range(0, len(Q)):
+                    H += ( Q[i][j] * qubo_basis[i] * qubo_basis[j])
+
+            for i in range(0, len(W)):
+                H+=(W[i]* qubo_basis[i])
+
+            #H+= float(bias) 
+    
+            #if self.verbose: print(H)
+            model = H.compile()
+            #if self.verbose: print(model)
+            bqm = model.to_bqm()
+            #if self.verbose: print(bqm)
+
+
+            # solve the model
+            sa = neal.SimulatedAnnealingSampler()
+
+            sampleset = sa.sample(bqm, num_sweeps =100, num_reads=10)
+            
+            #if self.verbose: print(sampleset)
+            decoded_samples = model.decode_sampleset(sampleset)
+            best_sample = min(decoded_samples, key=lambda x: x.energy)
+
+
+            return best_sample
+
+        
+        def get_qubo_hamiltonian_minimas(self, num_samples):
```

Comment:
Could get_qubo_hamiltonian_minimas and get_qubo_hamiltonian_minima be simplified into one method that determines if one or multiple minima should be computed based on user input?